tests/store/artifact/test_http_artifact_repo.py

```diff
@@ -206,38 +206,41 @@ def test_download_artifacts(http_artifact_repo, tmpdir):
     # - dir
     #   - b.txt
     # ---------
-    side_effect = [
-        # Response for `list_experiments("")` called by `_is_directory("")`
-        MockResponse(
-            {
-                "files": [
-                    {"path": "a.txt", "is_dir": False, "file_size": 6},
-                    {"path": "dir", "is_dir": True},
-                ]
-            },
-            200,
-        ),
-        # Response for `list_experiments("")`
-        MockResponse(
-            {
-                "files": [
-                    {"path": "a.txt", "is_dir": False, "file_size": 6},
-                    {"path": "dir", "is_dir": True},
-                ]
-            },
-            200,
-        ),
-        # Response for `_download_file("a.txt")`
-        MockStreamResponse("data_a", 200),
-        # Response for `list_experiments("dir")`
-        MockResponse({"files": [{"path": "b.txt", "is_dir": False, "file_size": 1}]}, 200),
```

Comment:
```
    def list_artifacts(self, path=None):
        endpoint = "/mlflow-artifacts/artifacts"
        url, tail = self.artifact_uri.split(endpoint, maxsplit=1)
        root = tail.lstrip("/")
        params = {"path": posixpath.join(root, path) if path else root}
        host_creds = _get_default_host_creds(url)
        resp = http_request(host_creds, endpoint, "GET", params=params)
        augmented_raise_for_status(resp)
        file_infos = []
>       for f in resp.json().get("files", []):
E       AttributeError: 'str' object has no attribute 'get'

endpoint   = '/mlflow-artifacts/artifacts'
file_infos = []
host_creds = <mlflow.utils.rest_utils.MlflowHostCreds object at 0x0000022673C56B88>
params     = {'path': 'dir'}
path       = 'dir'
resp       = <tests.store.artifact.test_mlflow_artifact_repo.MockStreamResponse object at 0x0000022671828D88>
root       = ''
self       = <mlflow.store.artifact.mlflow_artifacts_repo.MlflowArtifactsRepository object at 0x00000226718288C8>
tail       = ''
url        = 'http://localhost:5000/api/2.0'
```

Reply:
This error can be reproduced by:

```diff
diff --git a/mlflow/store/artifact/artifact_repo.py b/mlflow/store/artifact/artifact_repo.py
index 63a0195c3..a0c52e242 100644
--- a/mlflow/store/artifact/artifact_repo.py
+++ b/mlflow/store/artifact/artifact_repo.py
@@ -152,8 +152,14 @@ class ArtifactRepository:
             local_destination_file_path = self._create_download_destination(
                 src_artifact_path=src_artifact_path, dst_local_dir_path=dst_local_dir_path
             )
+            import time
+
+            def delayed_download_file(remote_file_path, local_path):
+                time.sleep(1)
+                return self._download_file(remote_file_path, local_path)
+
             download_future = self.thread_pool.submit(
-                self._download_file,
+                delayed_download_file,
                 remote_file_path=src_artifact_path,
                 local_path=local_destination_file_path,
             )
```

Reply:
This test assumes assumes the following request order:

```
1. list_artifacts
2. list_artifacts
3. download_artifact <-
4. list_artifacts.   <- 
5. download_artifact
```

but 3 and 4 could be swapped because an artifact is download asynchronously.

Reply:
In the fixed test, the request order doesn't matter.

Reply:
I've observed this error a few times before on Windows.